// @evershop/evershop/src/modules/tax/api/getGasPrice/getGasPrice.js
const { Pool } = require('pg');
const { OK, INTERNAL_SERVER_ERROR } = require('@evershop/evershop/src/lib/util/httpStatus');
const { select } = require('@evershop/postgres-query-builder');

const connectionSetting = {
  user: 'postgres',
  host: 'localhost',
  database: 'eventshop',
  password: 'chainlinkhackathon',
  port: 5432,
};

module.exports = async (request, response) => {
  const pool = new Pool(connectionSetting);

  try {
    await pool.connect();

    const gasPricesQuery = select('*').from('gas_price');

    let gasPrices  = await gasPricesQuery.execute(pool);

    if (!gasPrices || gasPrices.length === 0) {
      console.log('Brak danych z bazy lub dane są puste.');
    }

   

    response.status(OK).json(gasPrices);
  } catch (error) {
    console.error('Błąd podczas wykonywania zapytania', error);

    if (error.code) {
      console.error('Kod błędu:', error.code);
    }
    if (error.detail) {
      console.error('Szczegóły błędu:', error.detail);
    }
    if (error.hint) {
      console.error('Wskazówka błędu:', error.hint);
    }

    console.log('Zapytanie do bazy danych nie powiodło się.');

    response.status(INTERNAL_SERVER_ERROR).json({
      error: {
        status: INTERNAL_SERVER_ERROR,
        message: error.message
      }
    });
  }
};
