// @evershop/evershop/src/modules/tax/api/gasPrice/gasPrice_bak6.js
const {
  rollback,
  commit,
  startTransaction,
  update
} = require('@evershop/postgres-query-builder');
const { Pool } = require('pg');
const {
  OK,
  INTERNAL_SERVER_ERROR
} = require('@evershop/evershop/src/lib/util/httpStatus');

// Ustawienia połączenia do bazy danych
const connectionSetting = {
  user: 'postgres',
  host: 'localhost',
  database: 'eventshop',
  password: 'chainlinkhackathon',
  port: 5432,
};

module.exports = async (request, response, deledate, next) => {
  const pool = new Pool(connectionSetting);
  const connection = await pool.connect();
  await startTransaction(connection);
  
  const gasPrices = request.body; // Oczekujemy danych w formacie { 12: 333, 2: 0.013, ... }

  try {
    const updatedGasPrices = [];
    for (const key in gasPrices) {
      const id = parseInt(key); // Pobieramy identyfikator z klucza JSON
      const price = parseFloat(gasPrices[key].price);

      const gasPrice = await update('gas_price')
        .given({
          price // Ustawiamy nową cenę na podstawie przekazanej wartości
        })
        .where('id', '=', id) // Warunek WHERE dla identyfikatora
        .execute(connection);

      updatedGasPrices.push(gasPrice);
    }

    await commit(connection);
    response.status(OK);
    response.json({
      data: updatedGasPrices
    });
  } catch (e) {
    await rollback(connection);
    response.status(INTERNAL_SERVER_ERROR);
    response.json({
      error: {
        status: INTERNAL_SERVER_ERROR,
        message: e.message
      }
    });
  } finally {
    connection.release(); // Zwolnienie połączenia
  }
};
